### PR TITLE
Implement the replication flow controller

### PR DESCRIPTION
**Pull request by ywk253100**
1. Implement the replication flow controller
2. Implement the registry adapter management
3. Define the interface for PolicyManager, RegistryManager, ExecutionManager, TaskManager and Scheduler

Signed-off-by: Wenkai Yin <yinw@vmware.com>

### CODE REVIEW

File: src/replication/ng/adapter/adapter.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	// Fetch the content resource under the namespace by filters
+	// SUGGESTION: Adapter provider can do their own filter based on the filter pattern
+	// or call the default `DoFilter` function of the filter to complete resource filtering.
+	FetchResources(namespace string, filters []*model.Filter) ([]*model.Resource, error)
+}
+
+// Register one adapter to the registry, the param "adapter" is the type
+// of the adapter that you want to register
+func Register(name model.RegistryType, adapter interface{}) error {
+	if len(name) == 0 {
```
[the remark this answers is not in the document]
Comment (steven-zou) [reply src/replication/ng/adapter/adapter.go]: Not changed?

File: src/replication/ng/flow/flow.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+		return 0, fmt.Errorf("failed to create the execution record for replication based on policy %d: %v", policy.ID, err)
+	}
+	log.Debugf("an execution record for replication based on the policy %d created: %d", policy.ID, executionID)
+
+	// initialize
+	_, dstReg, srcAdapter, dstAdapter, err := initialize(policy, d.registryMgr)
+	if err != nil {
+		err = fmt.Errorf("failed to initialize the replication based on policy %d: %v", policy.ID, err)
+		log.Error(err.Error())
+		markExecutionFailure(d.executionMgr, executionID, err)
```
Comment (steven-zou) [inline src/replication/ng/flow/flow.go]: Can we define related state changing methods on the execution itself?

```go
func (e *execution) Failed() error{}

func (e *execution) Succeed() error{}

...
```

File: src/replication/ng/flow/flow.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+// defaultController is the default implement for the Controller
+type defaultController struct {
+	registryMgr  registry.Manager
+	executionMgr execution.Manager
+	scheduler    scheduler.Scheduler
+}
+
+// Replicate according the to policy ID
+func (d *defaultController) StartReplication(policy *model.Policy) (int64, error) {
```
Comment (steven-zou) [inline src/replication/ng/flow/flow.go]: Suggest to split the method into multi meaningful stages and then reduce the combined methods like `initialize` and `createNamespace` etc.. The body of this method is too large.

Comment (steven-zou) [reply src/replication/ng/flow/flow.go]: Can we create a flow object to cover those stages?

File: src/replication/ng/scheduler/scheduler.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"github.com/goharbor/harbor/src/replication/ng/model"
+)
+
+// Scheduler schedules tasks to transfer resource data
+type Scheduler interface {
+	// Schedule tasks for one execution
+	Schedule([]*model.Resource, []*model.Resource) ([]*model.Task, error)
+	// Stop the task specified by ID
+	StopTask(id string) error
```
Comment (steven-zou) [inline src/replication/ng/scheduler/scheduler.go]: Is it conflicted with PR https://github.com/goharbor/harbor/pull/6886? Can we use just `Stop` as the method name?

File: src/replication/ng/flow/controller.go [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+	StartReplication(policy *model.Policy) (int64, error)
+	// Stop the replication specified by the execution ID
+	StopReplication(int64) error
+}
+
+// NewController returns an instance of a Controller
+func NewController(registryMgr registry.Manager,
+	executionMgr execution.Manager, scheduler scheduler.Scheduler) (Controller, error) {
+	if registryMgr == nil || executionMgr == nil || scheduler == nil {
+		return nil, errors.New("invalid params")
```
Comment (steven-zou) [inline src/replication/ng/flow/controller.go]: The error is not accurate, seems a little general!

File: src/replication/ng/model/execution.go [hunk 6]
```diff
@@ -16,17 +16,38 @@ package model
...
-	Failed     int       `json:"failed"`
-	Succeed    int       `json:"succeed"`
-	Pending    int       `json:"pending"`
-	InProgress int       `json:"in_progress"`
-	StartTime  time.Time `json:"start_time"`
-	EndTime    time.Time `json:"end_time"`
+	ID         int64           `json:"id"`
+	PolicyID   int64           `json:"policy_id"`
+	Status     ExecutionStatus `json:"status"`
+	StatusText string          `json:"status_text"`
```
Comment (mmpei) [inline src/replication/ng/model/execution.go]: what Status and StatusText should be like?

File: src/replication/ng/model/execution.go [hunk 7]
```diff
@@ -16,17 +16,38 @@ package model
...
+	Total      int             `json:"total"`
+	Failed     int             `json:"failed"`
+	Succeed    int             `json:"succeed"`
+	Pending    int             `json:"pending"`
```
Comment (mmpei) [inline src/replication/ng/model/execution.go]: here should match the status defined above. Failed Succeed Stoped InProgress.
And i see there is a column called 'trigger_mode' in UI design.  i think we should store it in execution data if this value is mutable in policy.

File: src/replication/ng/model/execution.go [hunk 8]
```diff
@@ -37,7 +58,19 @@ type Task struct {
 	SrcResource  string       `json:"src_resource"`
 	DstResource  string       `json:"dst_resource"`
 	JobID        string       `json:"job_id"`
-	Status       string       `json:"status"`
+	Status       TaskStatus   `json:"status"`
```
Comment (mmpei) [inline src/replication/ng/model/execution.go]: define a type TaskStatus brings translation to dao.model and mananger.model and makes model could not be shared in dao and manager even they looks the same.

File: src/replication/ng/flow/flow.go [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+
+	tasks, err := f.scheduler.Schedule(f.resources, dstResources)
+	if err != nil {
+		f.markExecutionFailure(err)
+		return err
+	}
+
+	allFailed := true
+	for _, task := range tasks {
+		if task.Status != model.TaskStatusFailed {
```
Comment (mmpei) [inline src/replication/ng/flow/flow.go]: when a task will be set to failed even before creating?  does action here match with the one in scheduler?

File: src/replication/ng/model/execution.go [hunk 10]
```diff
@@ -16,17 +16,38 @@ package model
...
+const (
+	ExecutionStatusFailed     ExecutionStatus = "Failed"
+	ExecutionStatusSucceed    ExecutionStatus = "Succeed"
+	ExecutionStatusStopped    ExecutionStatus = "Stopped"
+	ExecutionStatusInProgress ExecutionStatus = "InProgress"
+	TaskStatusFailed          TaskStatus      = "Failed"
+	TaskStatusSucceed         TaskStatus      = "Succeed"
+	TaskStatusStopped         TaskStatus      = "Stopped"
+	TaskStatusInProgress      TaskStatus      = "InProgress"
+	TaskStatusPending         TaskStatus      = "Pending"
```
[the remark this answers is not in the document]
Comment (ywk253100) [reply src/replication/ng/model/execution.go]: No "Retrying" status, if a job need to be retry, it is put into the queue again, so it's status is "Pending". This is not supported by jobservice for now.

File: src/replication/ng/flow/flow.go (hunk 9 above)
Comment (ywk253100) [reply src/replication/ng/flow/flow.go]: The new changes that we talked about will be pushed in later PR.

File: src/replication/ng/model/execution.go [hunk 11]
```diff
@@ -41,3 +63,15 @@ type Task struct {
...
 	EndTime      time.Time    `json:"end_time"`
 }
+
+// ExecutionQuery defines the query conditions for listing executions
+type ExecutionQuery struct {
+	PolicyID int64
+}
+
+// TaskQuery defines the query conditions for listing tasks
+type TaskQuery struct {
```
Comment (cd1989) [inline src/replication/ng/model/execution.go]: We don't need pagination ?

**Comment by cd1989** [comment]
LGTM except one comment.

File: src/replication/ng/model/execution.go (hunk 11 above)
Comment (ywk253100) [reply src/replication/ng/model/execution.go]: Pagination added

**Pull request merged**